console/src/components/Services/Data/PermissionsSummary/RolesHeader.tsx

```diff
@@ -0,0 +1,71 @@
+import React from 'react';
+import Button from '../../../Common/Button/Button';
+import Header from './Header';
+import styles from './PermissionsSummary.scss';
+
+type RolesHeaderProps = {
+  selectable?: boolean;
+  selectedFirst?: boolean;
+  allRoles: Array<string>;
+  currentRole: string;
+  onCopyClick: (e: React.MouseEvent<HTMLButtonElement>, role: string) => void;
+  onDeleteClick: (e: React.MouseEvent<HTMLButtonElement>, role: string) => void;
+  setRole: (role: string, isCurrRole: boolean) => void;
+};
+
+type IconButtonProps = {
+  onClick: (e: React.MouseEvent<HTMLButtonElement>) => void;
+  icon: string;
+};
+
+const IconButton: React.FC<IconButtonProps> = ({ onClick, icon }) => (
+  <Button color="white" size="xs" onClick={onClick} title="Copy permissions">
+    <i className={`fa ${icon} ${styles.actionIcon}`} aria-hidden="true" />
+  </Button>
+);
+
+const RolesHeader: React.FC<RolesHeaderProps> = ({
+  selectable = true,
+  selectedFirst = false,
+  allRoles,
+  currentRole,
+  onCopyClick,
+  onDeleteClick,
+  setRole,
+}) => {
+  const rolesHeaders = [];
+
+  if (!allRoles.length) {
+    rolesHeaders.push(<Header content="No roles" selectable={false} />);
+  } else {
+    allRoles.forEach(role => {
+      const isCurrRole = currentRole === role;
+
+      const roleHeader = (
+        <Header
+          content={role}
+          selectable={selectable}
+          isSelected={isCurrRole}
+          onClick={() => setRole(role, isCurrRole)}
+          actionButtons={[
+            <IconButton icon="fa-copy" onClick={e => onCopyClick(e, role)} />,
+            <IconButton
+              icon="fa-trash"
+              onClick={e => onDeleteClick(e, role)}
+            />,
+          ]}
+        />
+      );
+
+      if (selectedFirst && isCurrRole) {
+        rolesHeaders.unshift(roleHeader);
+      } else {
+        rolesHeaders.push(roleHeader);
+      }
+    });
+  }
+
+  return <>{rolesHeaders}</>;
```

Comment:
So I really don't like this imperative style, so I was thinking about how we can make it more idiomatic react.

This part: `(selectedFirst && isCurrRole)` is responsible for positioning a role in the headers array and thus for the order of roles displayed in the UI. But we don't have to do it by `shift` and `unshift` of the components! We can order the array before doing any components work. 

How it can be done:
```ts
 let roles = allRoles;
  if (selectedFirst) {
    roles = allRoles.reduce((acc, role) => {
      if (role === currentRole) return [role, ...acc];
      return [...acc, role];
    }, [] as string[]);
  }
```

And then the component could look like this:

```tsx

const RolesHeader: React.FC<RolesHeaderProps> = ({
  selectable = true,
  selectedFirst = false,
  allRoles,
  currentRole,
  onCopyClick,
  onDeleteClick,
  setRole,
}) => {
  let roles = [...allRoles];
  if (selectedFirst) {
    roles = allRoles.reduce((acc, role) => {
      if (role === currentRole) return [role, ...acc];
      return [...acc, role];
    }, [] as string[]);
  }

  return (
    <>
      {roles.length ? roles.map(role => (
          <Header
            content={role}
            selectable={selectable}
            isSelected={currentRole === role}
            onClick={() => setRole(role, currentRole === role)}
            actionButtons={[
              <IconButton icon="fa-copy" onClick={e => onCopyClick(e, role)} />,
              <IconButton
                icon="fa-trash"
                onClick={e => onDeleteClick(e, role)}
              />,
            ]}
          />
        ))
      ) :  <Header content="No roles" selectable={false} />}
    </>
  );
};

export default RolesHeader;

```